vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,907 @@
+    {
+      // This is used to remove the label later.
+      data.action_with_label = (uint32_t)i;
+      attach_label_to_example(index, data.actions[i], slot->l.conditional_contextual_bandit.outcome, data);
+    }
+  }
+
+  // Must reset this in case the pooled example has stale data here.
+  data.shared->pred.a_s = v_init<ACTION_SCORE::action_score>();
+
+  // Tag can be used for specifying the sampling seed per slot. For it to be used it must be inserted into the shared example.
+  std::swap(data.shared->tag, slot->tag);
+}
+
+// iterate over slots contained in the multi-example, and for each slot, build a cb example and perform a
+// cb_explore_adf call.
+template <bool is_learn>
+void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
+{
+  clear_all(data);
+  split_multi_example(examples, data);  // split shared, actions and slots
+  sanity_checks<is_learn>(data);
+
+  // Stash the CCB labels before rewriting them.
+  data.stored_labels.clear();
+  for (auto ex : examples)
+  {
+    data.stored_labels.push_back({ex->l.conditional_contextual_bandit.type,
+        ex->l.conditional_contextual_bandit.outcome, ex->l.conditional_contextual_bandit.explicit_included_actions});
+  }
+
+  // Shared example must include history namespace:
+  if (std::find(data.shared->indices.begin(), data.shared->indices.end(), ccb_history_namespace) ==
+      data.shared->indices.end())
+  {
+    data.shared->indices.push_back(ccb_history_namespace);
+  }
+  data.shared->feature_space[ccb_history_namespace].push_back(1, constant);
+
+  // This will overwrite the labels with CB.
+  create_cb_labels(data);
+
+  // Reset exclusion list for this example.
+  data.exclude_list.assign(data.actions.size(), false);
+
+  auto decision_scores = v_init<ACTION_SCORE::action_scores>();
+
+  // for each slot, re-build the cb example and call cb_explore_adf
+  for (example* slot : data.slots)
+  {
+    multi_ex cb_ex;
+    build_cb_example<is_learn>(cb_ex, slot, data);
+
+    assert(std::find(data.shared->indices.begin(), data.shared->indices.end(), ccb_history_namespace) !=
+        data.shared->indices.end());
+
+    // Namespace crossing for slot features.
+    // If the slot example only has the constant namespace, there will be no extra crossing and so skip that logic.
+    if (!(slot->indices.size() == 1 && slot->indices[0] == constant_namespace))
+    {
+      data.generated_interactions.clear();
+      // TODO be more efficient here
+      std::copy(data.original_interactions->begin(), data.original_interactions->end(),
+          std::back_inserter(data.generated_interactions));
+      // TODO currently all action namespaces are used adding redundant interactions.
+      calculate_and_insert_interactions(data.shared, slot, data.actions, data.generated_interactions);
+      size_t removed_cnt;
+      size_t sorted_cnt;
+      INTERACTIONS::sort_and_filter_duplicate_interactions(data.generated_interactions, true, removed_cnt, sorted_cnt);
```

Comment:
Why do we need to call this here?  Can't we just guarantee that there are no duplicate interactions directly?

Reply:
This was a first pass attempt, the logic is simpler but less efficient. To do this without sort and filter, we would need to binary search to find if it exists and then insert if it does'nt. So it would be nlogn, unless we make a pass and build a map of indices then perhaps we could do it in n but with extra memory. I assume sort+filter is nlogn, so I figured it's not too bad to fall back to this way.